translator/trace/jaeger/protospan_to_jaegerproto.go

```diff
@@ -0,0 +1,477 @@
+			// base64 strings. Blindingly attempting to decode base64 seems too much.
+			str := truncableStringToStrProto(attribValue.StringValue)
+			jTag.VStr = str
+			jTag.VType = jaeger.ValueType_STRING
+		case *tracepb.AttributeValue_IntValue:
+			i := attribValue.IntValue
+			jTag.VInt64 = i
+			jTag.VType = jaeger.ValueType_INT64
+		case *tracepb.AttributeValue_BoolValue:
+			b := attribValue.BoolValue
+			jTag.VBool = b
+			jTag.VType = jaeger.ValueType_BOOL
+		case *tracepb.AttributeValue_DoubleValue:
+			d := attribValue.DoubleValue
+			jTag.VFloat64 = d
+			jTag.VType = jaeger.ValueType_FLOAT64
+		default:
+			str := "<Unknown OpenCensus Attribute for key \"" + key + "\">"
+			jTag.VStr = str
+			jTag.VType = jaeger.ValueType_STRING
+		}
+		jTags = append(jTags, jTag)
+	}
+
+	return jTags
+}
+
+func ocTimeEventsToJaegerLogsProto(ocSpanTimeEvents *tracepb.Span_TimeEvents) []jaeger.Log {
+	if ocSpanTimeEvents == nil || ocSpanTimeEvents.TimeEvent == nil {
+		return []jaeger.Log{}
+	}
+
+	ocTimeEvents := ocSpanTimeEvents.TimeEvent
+
+	// Assume that in general no time events are going to produce nil Jaeger logs.
+	jLogs := make([]jaeger.Log, 0, len(ocTimeEvents))
+	for _, ocTimeEvent := range ocTimeEvents {
+		jLog := jaeger.Log{
+			Timestamp: timestampToTimeProto(ocTimeEvent.Time),
+		}
+		switch teValue := ocTimeEvent.Value.(type) {
+		case *tracepb.Span_TimeEvent_Annotation_:
+			jLog.Fields = ocAnnotationToJagerTagsProto(teValue.Annotation)
+		case *tracepb.Span_TimeEvent_MessageEvent_:
+			jLog.Fields = ocMessageEventToJaegerTagsProto(teValue.MessageEvent)
+		default:
+			msg := "An unknown OpenCensus TimeEvent type was detected when translating to Jaeger"
+			jKV := jaeger.KeyValue{
+				Key:  ocTimeEventUnknownType,
+				VStr: msg,
+			}
+			jLog.Fields = append(jLog.Fields, jKV)
+		}
+
+		jLogs = append(jLogs, jLog)
+	}
+
+	return jLogs
+}
+
+func ocAnnotationToJagerTagsProto(annotation *tracepb.Span_TimeEvent_Annotation) []jaeger.KeyValue {
+	if annotation == nil {
+		return []jaeger.KeyValue{}
+	}
+
+	// TODO: Find a better tag for Annotation description.
+	jKV := jaeger.KeyValue{
+		Key:  ocTimeEventAnnotationDescription,
+		VStr: annotation.Description,
+	}
+	return append(ocSpanAttributesToJaegerTagsProto(annotation.Attributes), jKV)
+}
+
+func ocMessageEventToJaegerTagsProto(msgEvent *tracepb.Span_TimeEvent_MessageEvent) []jaeger.KeyValue {
+	if msgEvent == nil {
+		return []jaeger.KeyValue{}
+	}
+
+	msgEventID := make([]byte, 8)
+	binary.BigEndian.PutUint64(msgEventID, uint64(msgEvent.Id))
+
+	uncompressedSize := make([]byte, 8)
+	binary.BigEndian.PutUint64(uncompressedSize, uint64(msgEvent.UncompressedSize))
+
+	compressedSize := make([]byte, 8)
+	binary.BigEndian.PutUint64(compressedSize, uint64(msgEvent.CompressedSize))
+
+	// TODO: Find a better tag for Message event.
+	jaegerKVs := []jaeger.KeyValue{
+		jaeger.KeyValue{
+			Key:    ocTimeEventMessageEventType,
+			VInt64: msgEvent.Type,
```

Comment:
VSCode complains here (and a couple other places) with - 
```
cannot use msgEvent.Type (type "github.com/census-instrumentation/opencensus-proto/gen-go/trace/v1".Span_TimeEvent_MessageEvent_Type) as type int64 in field value
```
Should I add util functions for the conversions?

Reply:
Yea, I think it would be good.